provider/ec2/instancetypes.go

```diff
@@ -0,0 +1,328 @@
+	}
+	if clock == nil {
+		return vcpu * 100
+	}
+
+	// If the information includes a clock speed, we use that
+	// to estimate the ECUs. The pricing information does not
+	// include the ECUs, but they're only estimates anyway.
+	// Amazon moved to "vCPUs" quite some time ago.
+	// To date, info.ClockSpeed can have the form "Up to <float> GHz" or
+	// "<float> GHz", so look for a float match.
+	return uint64(*clock * 1.4 * 100 * float64(vcpu))
+}
+
+// virtType returns a virt type that matches our simplestrams metadata
+// rather than what the instance type actually reports.
+func virtType(info *ec2.InstanceTypeInfo) *string {
+	// Only very old instance types are restricted to paravirtual.
+	switch strings.SplitN(*info.InstanceType, ".", 2)[0] {
+	case "t1", "m1", "c1", "m2":
+		return aws.String("paravirtual")
+	}
+
+	return aws.String("hvm")
+}
+
+// supportsClassic reports whether the instance type with the given
+// name can be run in EC2-Classic.
+//
+// At the time of writing, we know that the following instance type
+// families support only VPC: C4, M4, P2, T2, X1. However, rather
+// than hard-coding that list, we assume that any new instance type
+// families support VPC only, and so we hard-code the inverse of the
+// list at the time of writing.
+//
+// See:
+//     http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-vpc.html#vpc-only-instance-types
+func supportsClassic(instanceType string) bool {
+	parts := strings.SplitN(instanceType, ".", 2)
+	if len(parts) < 2 {
+		return false
+	}
+	switch strings.ToLower(parts[0]) {
+	case
+		"c1", "c3",
+		"cc2",
+		"cg1",
+		"cr1",
+		"d2",
+		"g2",
+		"hi1",
+		"hs1",
+		"i2",
+		"m1", "m2", "m3",
+		"r3",
+		"t1":
+		return true
+	}
+	return false
+}
+
+var archNames = map[string]string{
+	"x86":     arch.I386,
+	"x86_64":  arch.AMD64,
+	"arm":     arch.ARM,
+	"aarch64": arch.ARM64,
+}
+
+func archName(in string) string {
+	if archName, ok := archNames[in]; ok {
+		return archName
+	}
+	return in
+}
+
+func (e *environ) supportedInstanceTypes(ec2Session ec2iface.EC2API, ctx context.ProviderCallContext) (result []instances.InstanceType, err error) {
+	e.instTypesMutex.Lock()
+	defer e.instTypesMutex.Unlock()
+
+	defer func() {
+		if err == nil {
+			if len(e.instTypes) == 0 {
+				logger.Tracef("got instance types:\n%#v", result)
+			}
+			e.instTypes = result
+		}
+	}()
+
+	// Use a cached copy if populated as it's mildly
+	// expensive to fetch each time.
+	if len(e.instTypes) > 0 {
+		return e.instTypes, nil
+	}
+
+	// First get all the zone names for the current region.
+	zoneResults, err := ec2Session.DescribeAvailabilityZones(&ec2.DescribeAvailabilityZonesInput{
+		Filters: []*ec2.Filter{{
+			Name:   aws.String("region-name"),
+			Values: []*string{aws.String(e.cloud.Region)},
+		}},
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var zoneNames []string
+	zoneFilter := &ec2.Filter{Name: aws.String("location")}
+	for _, z := range zoneResults.AvailabilityZones {
```

Comment:
This needs to support local zones like `us-west-2-lax-1a` or filter them out completely. Or if only local zones support an instance type, mark it as deprecated.

Reply:
I am not sure we care about the distinction. The only reason we get zone names is so that we can use them to know when to mark an instance as deprecated (terrible name but it precedes this work). If an instance type is not available across all zones (combination of az plus local), then deprecated = true.